### PR TITLE
Add QuantumNaturalGradient optimizer

### PR DESCRIPTION
**Pull request by EthanFeld**
## Summary
For issue #207 
Add a new `QuantumNaturalGradient` optimizer to `qiboml.models.optimizers` for generic parametrized circuits.

This implementation:
- follows the existing optimizer API style used by `ExactGeodesicTransportCG`
- supports both `"exp_val"` and generic callable losses
- computes the metric tensor using Qibo's QFIM implementation
- uses the Fubini-Study metric by dividing QFIM by 4
- applies a regularized linear solve for the natural-gradient update

## Changes

- added `QuantumNaturalGradient` in `src/qiboml/models/optimizers.py`
- added QNG docs to:
  - `doc/source/advanced/optimizers.rst`
  - `doc/source/api-reference/qiboml.models.rst`
- added tests in `tests/test_models_optimizers.py` covering:
  - VQE convergence on small RY+CZ ansatze
  - metric tensor correctness and symmetry
  - callable loss support on autodiff backends
  - expected failure on non-autodiff backend
  - invalid `loss_fn` / invalid Hamiltonian / missing Hamiltonian
  - sparse Hamiltonian path
  - callback + early stopping behavior
  - linear solve fallback path

## AI Disclosure
Codex 5.4 was utilized to help with programming, testing and documenting the added code

### CODE REVIEW

File: src/qiboml/models/optimizers.py [hunk 1]
```diff
@@ -1,20 +1,356 @@
...
+    def cast(self, array, dtype=None, *args, **kwargs):
+        if dtype is not None:
+            try:
+                dtype = getattr(self._backend, np.dtype(dtype).name)
+            except (AttributeError, TypeError):
+                pass
+        return self._backend.cast(array, dtype=dtype, *args, **kwargs)
+
+    def __getattr__(self, name):
+        return getattr(self._backend, name)
```
Comment (renatomello) [inline src/qiboml/models/optimizers.py]: none of this is necessary

File: src/qiboml/models/optimizers.py [hunk 2]
```diff
@@ -1,20 +1,356 @@
...
+        )
+        self.circuit.set_parameters(self.parameters)
+
+        if not isinstance(loss_fn, (str, Callable)):
+            raise_error(
+                TypeError,
+                "``loss_fn`` must be either the str ``exp_val`` or "
+                + f"``Callable``. Passed {type(loss_fn)}",
+            )
+        elif isinstance(loss_fn, str) and loss_fn != "exp_val":
```
Comment (renatomello) [inline src/qiboml/models/optimizers.py]:
```suggestion
        if isinstance(loss_fn, str) and loss_fn != "exp_val":
```

File: src/qiboml/models/optimizers.py [hunk 3]
```diff
@@ -1,20 +1,356 @@
...
+    ) -> Tuple[float, ArrayLike, ArrayLike]:
+        """Run QNG optimization loop."""
+
+        return self.run_qng(steps=steps, tolerance=tolerance)
+
+    def _loss_internal(self, circuit: Circuit, backend: Backend, **kwargs) -> float:
+        """Wrapper around loss function updating ``n_calls_loss``."""
+
+        self.n_calls_loss += 1
+        return self.loss_fn(circuit, backend, **kwargs)
```
Comment (renatomello) [inline src/qiboml/models/optimizers.py]: Seems unnecessary. No need for a whole method just to do `self.n_calls_loss += 1`

File: src/qiboml/models/optimizers.py [hunk 4]
```diff
@@ -1,20 +1,356 @@
...
+            return self.backend.to_numpy(
+                self.backend.engine.sparse.to_dense(self.hamiltonian)
+            )
+
+        if self.backend.platform == "pytorch" and getattr(
+            self.hamiltonian, "is_sparse", False
+        ):
+            return self.hamiltonian.to_dense().cpu().numpy(force=True)
+
+        return np.asarray(self.backend.to_numpy(self.hamiltonian))
```
Comment (renatomello) [inline src/qiboml/models/optimizers.py]: Not something that is supposed to be a method of an optimizer. Moreover, `hamiltonian` representation should not be converted to `numpy` without any good reason.

File: tests/test_models_optimizers.py [hunk 5]
```diff
@@ -1,13 +1,44 @@
...
+    try:
+        _ = _get_autodiff_backend(_backend_name)
+        AUTODIFF_BACKENDS.append(_backend_name)
+    except (ModuleNotFoundError, ImportError):
+        pass
+
+
+@pytest.fixture(params=AUTODIFF_BACKENDS)
+def autodiff_backend(request):
+    yield _get_autodiff_backend(request.param)
```
Comment (renatomello) [inline tests/test_models_optimizers.py]: None of this is necessary.

File: tests/test_models_optimizers.py [hunk 6]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+    for qubit in range(nqubits - 1):
+        circuit.add(gates.CZ(qubit, qubit + 1))
+    for qubit in range(nqubits):
+        circuit.add(gates.RY(qubit, theta=0.0))
+    return circuit
+
+
+@pytest.mark.parametrize("nqubits", [2, 3])
+def test_quantum_natural_gradient_vqe(autodiff_backend, nqubits):
+    backend = autodiff_backend
```
Comment (renatomello) [inline tests/test_models_optimizers.py]:
```suggestion
@pytest.mark.parametrize("nqubits", [2, 3])
def test_quantum_natural_gradient_vqe(backend, nqubits):
```

File: src/qiboml/models/optimizers.py [hunk 7]
```diff
@@ -1,20 +1,356 @@
...
+        learning_rate: float = 0.01,
+        regularization: float = 1e-3,
+        callback: Callable[..., None] | None = None,
+        seed: int | None = None,
+        backend: Backend = None,
+    ):
+        del seed  # API compatibility; initialization comes from circuit parameters.
+
+        self.backend = _check_backend(backend)
+        self.qibo_backend = _BackendCastAdapter(self.backend)
```
Comment (renatomello) [inline src/qiboml/models/optimizers.py]:
```suggestion
```

File: src/qiboml/models/optimizers.py [hunk 8]
```diff
@@ -1,20 +1,356 @@
...
+                raise_error(
+                    ValueError,
+                    "For ``loss_fn='exp_val'``, you must pass the hamiltonian to ``loss_kwargs`` "
+                    + "via the dict item ``{'hamiltonian': hamiltonian}``",
+                )
+            self.loss_fn = _loss_func_expval
+            self.parameter_shift_hamiltonian = Hamiltonian(
+                self.circuit.nqubits,
+                self._hamiltonian_to_dense(),
+                backend=self.qibo_backend,
```
Comment (renatomello) [inline src/qiboml/models/optimizers.py]:
```suggestion
                backend=self.backend,
```

File: src/qiboml/models/optimizers.py [hunk 9]
```diff
@@ -1,20 +1,356 @@
...
+
+        self.loss = self._loss_internal
+
+    def metric_tensor(self) -> ArrayLike:
+        """Compute Fubini-Study metric tensor for current circuit parameters."""
+
+        qfim = quantum_fisher_information_matrix(
+            self.circuit,
+            parameters=self.parameters,
+            backend=self.qibo_backend,
```
Comment (renatomello) [inline src/qiboml/models/optimizers.py]:
```suggestion
            backend=self.backend,
```

File: tests/test_models_optimizers.py [hunk 10]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+        circuit=circuit,
+        loss_fn="exp_val",
+        loss_kwargs={"hamiltonian": hamiltonians.Z(3, backend=backend).matrix},
+        backend=backend,
+    )
+    metric = optimizer.metric_tensor()
+    qfim = quantum_fisher_information_matrix(
+        circuit,
+        parameters=parameters,
+        backend=optimizer.qibo_backend,
```
Comment (renatomello) [inline tests/test_models_optimizers.py]:
```suggestion
        backend=backend,
```

File: tests/test_models_optimizers.py [hunk 11]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+    )
+
+    assert optimizer.n_calls_loss >= len(losses)
+    assert optimizer.n_calls_gradient >= 1
+    backend.assert_allclose(final_parameters, circuit_parameters, atol=1e-7)
+    assert backend.abs(final_loss + nqubits) < 1e-5
+
+
+def test_quantum_natural_gradient_metric_tensor(autodiff_backend):
+    backend = autodiff_backend
```
Comment (renatomello) [inline tests/test_models_optimizers.py]:
```suggestion
def test_quantum_natural_gradient_metric_tensor(backend):
```

File: tests/test_models_optimizers.py [hunk 12]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+        loss_kwargs={"hamiltonian": hamiltonians.Z(3, backend=backend).matrix},
+        backend=backend,
+    )
+    metric = optimizer.metric_tensor()
+    qfim = quantum_fisher_information_matrix(
+        circuit,
+        parameters=parameters,
+        backend=optimizer.qibo_backend,
+    )
+    expected = backend.cast(backend.real(qfim) / 4.0, dtype=backend.float64)
```
Comment (renatomello) [inline tests/test_models_optimizers.py]:
```suggestion
    expected = backend.real(qfim) / 4.0
```

File: tests/test_models_optimizers.py [hunk 13]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+    metric_transpose = backend.cast(
+        np.transpose(backend.to_numpy(metric)),
+        dtype=backend.float64,
+    )
```
Comment (renatomello) [inline tests/test_models_optimizers.py]:
```suggestion
    metric_transpose = backend.transpose(metric)
```

File: tests/test_models_optimizers.py [hunk 14]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+
+
+def test_quantum_natural_gradient_callable_loss(autodiff_backend):
+    backend = autodiff_backend
+    circuit = _build_ry_cz_ansatz(2)
+    initial_parameters = backend.cast([0.2, -0.1, 0.3, 0.4], dtype=backend.float64)
+
+    def callable_loss(circuit, backend):
+        state = backend.execute_circuit(circuit).state()
+        return backend.real(state[0] * backend.conj(state[0]))
```
Comment (renatomello) [inline tests/test_models_optimizers.py]:
```suggestion
        return backend.real(state * backend.conj(state))
```

Comment (renatomello) [reply tests/test_models_optimizers.py]: This loss is not a real loss function that makes sense to the problem that is being solved.

File: tests/test_models_optimizers.py [hunk 15]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+        initial_parameters=initial_parameters,
+        learning_rate=0.05,
+        regularization=1e-3,
+        backend=backend,
+    )
+    loss = optimizer.step()
+
+    assert optimizer.n_calls_gradient == 1
+    assert optimizer.n_calls_loss >= 1
+    assert np.isfinite(np.asarray(backend.to_numpy(loss), dtype=np.float64)).all()
```
Comment (renatomello) [inline tests/test_models_optimizers.py]: this test makes no sense

File: tests/test_models_optimizers.py [hunk 16]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+
+
+def test_quantum_natural_gradient_callable_loss_errors():
+    backend = NumpyBackend()
+    circuit = _build_ry_cz_ansatz(2)
+
+    def callable_loss(circuit, backend):
+        state = backend.execute_circuit(circuit).state()
+        return backend.real(state[0] * backend.conj(state[0]))
```
Comment (renatomello) [inline tests/test_models_optimizers.py]:
```suggestion
    def callable_loss(circuit, backend):
        state = backend.execute_circuit(circuit).state()
        return backend.real(state * backend.conj(state))
```

Same issue as above

File: tests/test_models_optimizers.py [hunk 17]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+        _ = QuantumNaturalGradient(
+            circuit=circuit,
+            loss_fn=callable_loss,
+            backend=backend,
+        )
+
+
+def test_quantum_natural_gradient_errors(autodiff_backend):
+    backend = autodiff_backend
+    circuit = _build_ry_cz_ansatz(2)
```
Comment (renatomello) [inline tests/test_models_optimizers.py:420]:
```suggestion
def test_quantum_natural_gradient_errors(backend):
```

File: tests/test_models_optimizers.py [hunk 18]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+            loss_fn="exp_val",
+            loss_kwargs={},
+            backend=backend,
+        )
+
+
+def test_quantum_natural_gradient_sparse_hamiltonian_callback_and_tolerance(
+    autodiff_backend,
+):
+    backend = autodiff_backend
```
Comment (renatomello) [inline tests/test_models_optimizers.py]:
```suggestion
def test_quantum_natural_gradient_sparse_hamiltonian_callback_and_tolerance(
    backend,
):
```

File: tests/test_models_optimizers.py [hunk 19]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+    optimizer.gradient = lambda: backend.cast(
+        np.zeros(len(optimizer.parameters)),
+        dtype=backend.float64,
+    )
+
+    final_loss, losses, _ = optimizer.run_qng(steps=5, tolerance=1e-12)
+
+    assert len(losses) == 1
+    assert len(callback_calls) == 1
+    assert np.isfinite(np.asarray(backend.to_numpy(final_loss), dtype=np.float64)).all()
```
Comment (renatomello) [inline tests/test_models_optimizers.py]: Same problem as above

File: tests/test_models_optimizers.py [hunk 20]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+
+
+def test_quantum_natural_gradient_linear_solve_fallback(autodiff_backend, monkeypatch):
+    backend = autodiff_backend
```
Comment (renatomello) [inline tests/test_models_optimizers.py]:
```suggestion
def test_quantum_natural_gradient_linear_solve_fallback(backend):
```

File: tests/test_models_optimizers.py [hunk 21]
```diff
@@ -274,3 +305,202 @@ def _get_xxz_hamiltonian(nqubits, hamiltonian_type, backend):
...
+        backend=backend,
+    )
+
+    def fail_solve(*args, **kwargs):
+        raise np.linalg.LinAlgError()
+
+    monkeypatch.setattr(np.linalg, "solve", fail_solve)
+    loss = optimizer.step()
+
+    assert np.isfinite(np.asarray(backend.to_numpy(loss), dtype=np.float64)).all()
```
Comment (renatomello) [inline tests/test_models_optimizers.py]: This is not a real or meaningful test

**Comment by EthanFeld** [comment]
@renatomello thank you for the feedback, I implemented it + some other bug fixes in the new commits
